homeassistant/components/nest/camera.py

```diff
@@ -79,39 +83,71 @@ def __init__(self, device: Device) -> None:
         self._attr_device_info = nest_device_info.device_info
         self._attr_brand = nest_device_info.device_brand
         self._attr_model = nest_device_info.device_model
-        self._rtsp_stream: RtspStream | None = None
-        self._webrtc_sessions: dict[str, WebRtcStream] = {}
-        self._create_stream_url_lock = asyncio.Lock()
-        self._stream_refresh_unsub: Callable[[], None] | None = None
-        self._attr_is_streaming = False
-        self._attr_supported_features = CameraEntityFeature(0)
-        self._rtsp_live_stream_trait: CameraLiveStreamTrait | None = None
-        if CameraLiveStreamTrait.NAME in self._device.traits:
-            self._attr_is_streaming = True
-            self._attr_supported_features |= CameraEntityFeature.STREAM
-            trait = cast(
-                CameraLiveStreamTrait, self._device.traits[CameraLiveStreamTrait.NAME]
-            )
-            if StreamingProtocol.RTSP in trait.supported_protocols:
-                self._rtsp_live_stream_trait = trait
         self.stream_options[CONF_EXTRA_PART_WAIT_TIME] = 3
         # The API "name" field is a unique device identifier.
         self._attr_unique_id = f"{self._device.name}-camera"
+        self._stream_refresh_unsub: Callable[[], None] | None = None
 
-    @property
-    def use_stream_for_stills(self) -> bool:
-        """Whether or not to use stream to generate stills."""
-        return self._rtsp_live_stream_trait is not None
+    @abstractmethod
+    def _stream_expires_at(self) -> datetime.datetime | None:
+        """Next time when a stream expires."""
+
+    @abstractmethod
+    async def _async_refresh_stream(self) -> None:
+        """Refresh any stream to extend expiration time."""
+
+    def _schedule_stream_refresh(self) -> None:
+        """Schedules an alarm to refresh any streams before expiration."""
+        if self._stream_refresh_unsub is not None:
+            self._stream_refresh_unsub()
+
+        expiration_time = self._stream_expires_at()
+        if not expiration_time:
+            return
+        refresh_time = expiration_time - STREAM_EXPIRATION_BUFFER
+        _LOGGER.debug("Scheduled next stream refresh for %s", refresh_time)
+
+        self._stream_refresh_unsub = async_track_point_in_utc_time(
+            self.hass,
+            self._handle_stream_refresh,
+            refresh_time,
+        )
+
+    async def _handle_stream_refresh(self, _: datetime.datetime) -> None:
+        """Alarm that fires to check if the stream should be refreshed."""
+        _LOGGER.debug("Examining streams to refresh")
+        await self._async_refresh_stream()
+        self._schedule_stream_refresh()
```

Comment:
Should this be put in a `finally` block to ensure a refresh is always scheduled, even if a connection fails?